lib/puppet/type/agent_upgrade.rb

```diff
@@ -0,0 +1,30 @@
+Puppet::Type.newtype(:agent_upgrade) do
+  ensurable
+
+  newparam(:name, :namevar => true)
+
+  newparam(:source) do
+    newvalues /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+    validate do |value|
+      if value !~ /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+        fail "Please provide a valid http(s), puppet or unc path"
+      end
+      if value =~ /^https?/ && value !~ URI::regexp
+        fail("Please provide a valid URI")
```

Comment:
Interesting, so we are stating we know all the ways the user may choose to upgrade and we are going to limit them. I'd almost prefer to let the user decide, but check what they are attempting to install to be sure it is not an older version of Puppet.